### PR TITLE
[Persistence] TxIndexer to StateHash Bridge (Issue-#315)

### LINKED ISSUE #315: [Persistence] TxIndexer to StateHash Bridge
## Objective

The objective of this issue is to "bridge" the work being done in #302 (tx indexer integration) and #285 (first state hash implementation.

## Origin Document

The interface and diagrams for the state hash computation were being done in #252 while the first implementation began in #285.  However, since both of these are dependent on the integration of the tx indexer in #302, some conflicting design decisions arose. 

The goal of this ticket is to unblock #302, while also capturing the results of an offline discussion that led to a decision where the "ephemeral block state" should be wholly managed by the persistence context, being a single, "roll-backable" point of reference.

## Goals

- [ ] Remove ephemeral block-related state from the consensus module & utility module, and only maintain it in the persistence module
- [ ] Simply outward-facing module interfaces to only rely on primitive types
- [ ] Enable committing of a block to persistence by simply calling `Commit(quorumCert)`; see #252 for more details

After #302, #285 and this ticket is done, we should have a design that more closely resembles the following
* P2P - Wakes up consensus module (pipe for consensus module)
* Consensus - State Machine Orchestration
* Utility module - State Machine
* Postgres Context - Maintains state

## Deliverable

A PR that achieves the goals and includes, but not limited to, the following changes:
- [ ] Remove `lastAppHash` from the `consensusModule` struct state & access it from the persistence layer (added n the past)
- [ ] Remove `validatorMap` from the `consensusModule` struct state & access it from the persistence layer (added n the past)
- [ ] Remove `TxResult` from the utility module interface (added in #302)
- [ ] Encapsulate TxIndexer logic only inside the persistence context

## Non-goals / Non-deliverables

- Implementation of the state commitment (i.e. state hash)

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## Testing Methodology

- Add new unit tests if applicable
- **All tests**: `make develop_test`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @Olshansk 
**Co-Owners**: @andrewnguyen22

### PR DESCRIPTION
**Pull request by andrewnguyen22**
## Description

**TL;DR Encapsulate and commit the block and block parts within a persistence context**

The objective of this issue is to "bridge" the work being done in https://github.com/pokt-network/pocket/pull/302 (tx indexer integration) and https://github.com/pokt-network/pocket/pull/285 (first state hash implementation.

The interface and diagrams for the state hash computation were being done in https://github.com/pokt-network/pocket/pull/252 while the first implementation began in https://github.com/pokt-network/pocket/pull/285. However, since both of these are dependent on the integration of the tx indexer in https://github.com/pokt-network/pocket/pull/302, some conflicting design decisions arose.

The goal of this ticket is to unblock https://github.com/pokt-network/pocket/pull/302, while also capturing the results of an offline discussion that led to a decision where the "ephemeral block state" should be wholly managed by the persistence context, being a single, "roll-backable" point of reference

- [x]  **Enable committing of a block to persistence by simply committing the persistence context**
- [x] Remove ephemeral block-related state from the consensus module & utility module, and only maintain it in the persistence module
- [x]  Simply outward-facing module interfaces to only rely on primitive types
- [ ] Remove validatorMap from the consensusModule struct state & access it from the persistence layer (added n the past) 

Opened new issue for last one to prevent scope from getting out of hand: https://github.com/pokt-network/pocket/issues/331

## Issue

Fixes #315 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Removed `apphash` and `txResults` from `consensus Module` structure
- Modified lifecycle to `set` the proposal block within a Persistence Context 
- Allow block and parts to be committed with the persistence context
- Ported over storing blocks and block parts to persistence module from Consensus and Utility
- Encapsulate TxIndexer logic only inside the persistence context
- Remove TxResult from the utility module interface (added in [TxIndexer] Integration of transaction indexer (issue-#168) #302)
- Combined creation and application of block in proposer lifecycle

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: shared/modules/persistence_module.go [hunk 1]
```diff
@@ -55,15 +55,12 @@ type PersistenceWriteContext interface {
...
-	StoreTransaction(txResult TxResult) error
+	StoreTransactions() error
 
 	// Block Operations
-	// TEMPORARY: Including two functions for the SQL and KV Store as an interim solution
-	//                 until we include the schema as part of the SQL Store because persistence
-	//                 currently has no access to the protobuf schema which is the source of truth.
-	StoreBlock(blockProtoBytes []byte) error                                              // Store the block in the KV Store
-	InsertBlock(height uint64, hash string, proposerAddr []byte, quorumCert []byte) error // Writes the block in the SQL database
+	SetLatestTxResults(txResults []TxResult)
```
Comment (Olshansk) [inline shared/modules/persistence_module.go:61]: I'm beginning to lose context of what `last` and `latest` mean throughout the codebase, and understand I also added to this confusion in different parts of the code.

Do you think it makes sense to use `SetProposalTxResults` here instead so it's clear it is the current proposal and not yet committed? Ditto for the primitives backing this.

File: persistence/block.go [hunk 2]
```diff
@@ -55,15 +76,51 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
...
+		}
+	}
+	return nil
 }
 
-func (p PostgresContext) StoreBlock(blockProtoBytes []byte) error {
+// DISCUSS: this might be retrieved from the block store - temporarily we will access it directly from the module
+//       following the pattern of the Consensus Module prior to pocket/issue-#315
+func (p *PostgresContext) SetProposalBlock(blockHash string, blockProtoBytes, proposerAddr, qc []byte, transactions [][]byte) error {
+	p.latestBlockHash = blockHash
```
Comment (Olshansk) [inline persistence/block.go:92]: Per my comment above, thinking of renaming these to `proposalX` instead of `latestX` since it is only a proposal until the context is commited.

File: persistence/block.go [hunk 3]
```diff
@@ -55,15 +76,51 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
...
+	p.latestBlockProtoBytes = blockProtoBytes
```
Comment (Olshansk) [inline persistence/block.go:93]: The `blockProtoBytes` are a serialization of the consensus block proto. I don't think we need this. The interface transfers the primitives (address, QC, etc...) from consensus to persistence, and the bytes that'll be stored are those that are a serialization of the persistence specific proto (which I add in the follow up commit).

File: persistence/block.go [hunk 4]
```diff
@@ -39,6 +40,26 @@ func (p PostgresContext) GetHeight() (int64, error) {
 	return p.Height, nil
 }
 
+func (p PostgresContext) GetLastAppHash() (string, error) {
```
Comment (Olshansk) [inline persistence/block.go:43]: Thoughts on renaming this to `GetPrevAppHash`?

File: persistence/block.go [hunk 5]
```diff
@@ -39,6 +40,26 @@ func (p PostgresContext) GetHeight() (int64, error) {
...
+		return "", err
+	}
+	if height <= 1 {
+		return "TODO: get from genesis", nil
+	}
+	block, err := p.blockstore.Get(heightToBytes(height - 1))
+	if err != nil {
+		return "", fmt.Errorf("error getting block hash for height %d even though it's in the database: %s", height, err)
+	}
+	// IMPROVE/CLEANUP: returning the whole protoblock - should just return hash
```
Comment (Olshansk) [inline persistence/block.go:55]:
```suggestion
	// IMPROVE/CLEANUP(#284): returning the whole protoblock - should just return hash
```

File: persistence/block.go [hunk 6]
```diff
@@ -55,15 +76,51 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
 	return true, err
 }
 
-func (p PostgresContext) StoreTransaction(txResult modules.TxResult) error {
-	return p.txIndexer.Index(txResult)
+func (p PostgresContext) StoreTransactions() error {
```
Comment (Olshansk) [inline persistence/block.go:79]: I think `IndexProposalTransactions` would be more representative of what this does.

File: persistence/db.go [hunk 7]
```diff
@@ -43,6 +43,22 @@ type PostgresContext struct {
 	tx         pgx.Tx
 	blockstore kvstore.KVStore
 	txIndexer  indexer.TxIndexer
+	// DISCUSS: this might be retrieved from the block store - temporarily we will access it directly from the module
```
Comment (Olshansk) [inline persistence/db.go:46]: I think this is more than fine for now. Let's merge it in because I want to get "a feel" for how we can do this well after I rebase.

```suggestion
	// DISCUSS(#284): this might be retrieved from the block store - temporarily we will access it directly from the module
```

File: persistence/context.go [hunk 8]
```diff
@@ -48,6 +57,9 @@ func (p PostgresContext) Release() error {
 	if err := p.conn.Close(ctx); err != nil {
 		log.Println("[TODO][ERROR] Implement connection pooling. Error when closing DB connecting...", err)
 	}
+	if err := p.Reset(); err != nil {
```
Comment (Olshansk) [inline persistence/context.go:60]: We should also call Reset at the end of `Commit`

File: persistence/block.go [hunk 9]
```diff
@@ -55,15 +76,51 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
...
+}
+
+// TEMPORARY: Including two functions for the SQL and KV Store as an interim solution
+//                 until we include the schema as part of the SQL Store because persistence
+//                 currently has no access to the protobuf schema which is the source of truth.
+// TODO: atomic operations needed here - inherited pattern from consensus module
+func (p PostgresContext) StoreBlock() error {
+	if p.latestBlockProtoBytes == nil {
+		// IMPROVE/CLEANUP: HACK - currently tests call Commit() on the same height and it throws a
+		// ERROR: duplicate key value violates unique constraint "block_pkey", because it attempts to
```
Comment (Olshansk) [inline persistence/block.go:107]: I already have this done in the next PR. Code and utilities :)

**Review by Olshansk — CHANGES_REQUESTED** [review]
**tl;dr I think we can get this in relatively quickly.**

1. I pushed some minor NITS in the changelog files; everything else left as a comment.

2. Business logic-wise, I don't have any real questions/comments. It makes sense and does the job of "bridging" what I needed in the next PR.

3. My only major point of discussion (and there's lots of comments related to it) is naming. I want to bias to using "Proposal" instead of "Latest" and "Previous" instead of "Last" where appropriate. Lmk what you think

File: utility/actor.go [hunk 10]
```diff
@@ -518,6 +518,11 @@ func (u *UtilityContext) CheckBelowMaxChains(actorType typesUtil.ActorType, chai
 	return nil
 }
 
+func (u *UtilityContext) GetLastBlockByzantineValidators() ([][]byte, error) {
+	// TODO: Need to retrieve byzantine validators from the persistence module
```
[the remark this answers is not in the document]
Comment (deblasis) [reply utility/actor.go:522]: I was just looking at how `ApplyBlock` in `Utility` changed and at how we are not passing in anymore the addresses of the `ByzantineValidators`, I was about to ask and I bumped into this comment :)

Who is in charge in determining who are the validators that disagreed with the 2/3+ majority or didn't sign at all?
It doesn't sound very "P2Py" or "Persistencey" to me 🤔

Sure, `Persistence` is the repository for them.
`P2P` to me sounds more specific around the communication side of things. 
`Consensus` maybe is the module that "knows" who the bad boys are.

Correct me if I am wrong.

[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/actor.go:522]: `Consensus`, `Utility` and `P2P` will all be reading the same set of validators from the persistence for a specific height, which is why we're not passing them around. I can see an approach to argue both ways by making things super functional and passing all the vals around, but we do have a pretty stateful lifecycle that's persistence reliant at the moment.

> Consensus maybe is the module that "knows" who the bad boys are.

That is correct.

That logic is currently spread across https://github.com/pokt-network/pocket/blob/main/consensus/hotstuff_replica.go#L265

Screen Shot 2022-11-02 at 6 43 36 PM

And https://github.com/pokt-network/pocket/blob/main/consensus/helpers.go#L118:

Screen Shot 2022-11-02 at 6 44 25 PM

---

We also talked in a protocol hour about how we will handle / punish double signing:

https://www.notion.so/pocketnetwork/2022-08-02-7c748d2d44434832a29027133ad94fed

Note: if you are an external contributor somehow reading this and would like access to the link above, please try to get in touch with the team :)

File: shared/modules/utility_module.go [hunk 11]
```diff
@@ -15,9 +15,8 @@ type UtilityModule interface {
 // operations.
 type UtilityContext interface {
 	// Block operations
-	// TODO(#315): Remove `txResults TxResult` from the interface
-	GetProposalTransactions(proposer []byte, maxTransactionBytes int, lastBlockByzantineValidators [][]byte) (transactions [][]byte, txResults []TxResult, err error)
-	ApplyBlock(height int64, proposer []byte, transactions [][]byte, lastBlockByzantineValidators [][]byte) (appHash []byte, txResults []TxResult, err error)
+	CreateAndApplyBlock(proposer []byte, maxTransactionBytes int) (appHash []byte, transactions [][]byte, err error)
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply shared/modules/utility_module.go:18]: Will do for `CreateAndApplyProposalBlock` but `ApplyBlock` will be used in state sync - with regular blocks. Will leave a comment

File: utility/actor.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/actor.go:522]: TL;DR - Consensus module determines byzantine validators - Persistence module stores it for future reads

File: persistence/block.go [hunk 12]
```diff
@@ -55,15 +76,51 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
 	return true, err
 }
 
-func (p PostgresContext) StoreTransaction(txResult modules.TxResult) error {
-	return p.txIndexer.Index(txResult)
+func (p PostgresContext) StoreTransactions() error {
+	// TODO: store in batch
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply persistence/block.go:80]: So this is a `TODO` to me - because it's mission critical for the application. We need atomic storage in a batch

File: utility/doc/CHANGELOG.md [hunk 13]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.0.0.8] - 2022-11-01
+- Remove TxResult from the utility module interface (added in [TxIndexer] Integration of transaction indexer (issue-#168) #302)
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/doc/CHANGELOG.md]: Oh it's a C/P error from github; will fix

File: shared/modules/persistence_module.go (hunk 1 above)
Comment (andrewnguyen22) [reply shared/modules/persistence_module.go:61]: > I'm beginning to lose context of what last and latest mean throughout the codebase, and understand I also added to this confusion in different parts of the code.

Understood - will take your suggestion below

> Do you think it makes sense to use SetProposalTxResults here instead so it's clear it is the current proposal and not yet committed? Ditto for the primitives backing this.

See comment above about `ApplyBlock` being used during state sync

File: persistence/block.go (hunk 2 above)
Comment (andrewnguyen22) [reply persistence/block.go:92]: See comment above about `ApplyBlock` being used during state sync

File: persistence/block.go (hunk 3 above)
Comment (andrewnguyen22) [reply persistence/block.go:93]: Okay - so what do you suggest? To me, since your comment is based on `future work` it might make sense to amend in that commit. 

```
if err := p.blockstore.Put(heightToBytes(p.Height), p.latestBlockProtoBytes); err != nil {
		return err
	}
``` 

We need to serialize and store for `make compose_and_watch` to work in `this commit`

File: persistence/block.go (hunk 6 above)
Comment (andrewnguyen22) [reply persistence/block.go:79]: See comment above about ApplyBlock being used during state sync

File: persistence/block.go (hunk 9 above)
Comment (andrewnguyen22) [reply persistence/block.go:107]: Awesome :)

File: persistence/block.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/block.go:80]: Makes sense. Leave as is then.

File: consensus/consensus_tests/utils_test.go [hunk 14]
```diff
@@ -359,22 +359,23 @@ func baseUtilityContextMock(t *testing.T) *modulesMock.MockUtilityContext {
 	ctrl := gomock.NewController(t)
 	utilityContextMock := modulesMock.NewMockUtilityContext(ctrl)
 	persistenceContextMock := modulesMock.NewMockPersistenceRWContext(ctrl)
+	persistenceContextMock.EXPECT().SetProposalBlock(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	persistenceContextMock.EXPECT().GetLastAppHash().Return("", nil).AnyTimes()
```
Comment (Olshansk) [inline consensus/consensus_tests/utils_test.go:363]: Failing tests, please look into this.

File: consensus/hotstuff_leader.go [hunk 15]
```diff
@@ -337,36 +334,33 @@ func (m *consensusModule) tempIndexHotstuffMessage(msg *typesCons.HotstuffMessag
...
-		return nil, nil, err
+		return nil, err
 	}
 
-	// OPTIMIZE: Determine if we can avoid the `ApplyBlock` call here
-	// Apply all the transactions in the block
-	appHash, txResults, err := m.utilityContext.ApplyBlock(int64(m.Height), m.privateKey.Address(), txs, lastByzValidators)
+	persistenceContext := m.utilityContext.GetPersistenceContext()
+
+	lastAppHash, err := persistenceContext.GetPrevAppHash()
```
Comment (Olshansk) [inline consensus/hotstuff_leader.go:353]: s/lastAppHash/prevAppHash

File: shared/modules/persistence_module.go [hunk 16]
```diff
@@ -55,15 +55,12 @@ type PersistenceWriteContext interface {
...
+	SetProposalBlock(blockHash string, blockProtoBytes, proposerAddr, qc []byte, transactions [][]byte) error
```
Comment (Olshansk) [inline shared/modules/persistence_module.go:62]: Add a `TODO(#284): Remove `blockProtoBytes` from the interface

File: persistence/block.go (hunk 3 above)
Comment (Olshansk) [reply persistence/block.go:93]: Okay understood. Adding a comment about adding a TODO

File: utility/test/transaction_test.go [hunk 17]
```diff
@@ -86,18 +86,18 @@ func TestUtilityContext_GetSignerCandidates(t *testing.T) {
...
 	txBz, err := tx.Bytes()
 	require.NoError(t, err)
 	require.NoError(t, ctx.CheckTransaction(txBz))
-	txs, txResults, er := ctx.GetProposalTransactions([]byte(proposer.GetAddress()), 10000, nil)
+	appHash, txs, er := ctx.CreateAndApplyProposalBlock([]byte(proposer.GetAddress()), 10000)
 	require.NoError(t, er)
 	require.Equal(t, len(txs), 1)
 	require.Equal(t, txs[0], txBz)
-	requireValidTestingTxResults(t, tx, txResults)
+	require.NotNilf(t, appHash, "")
```
Comment (Olshansk) [inline utility/test/transaction_test.go:100]: I think you can switch `NotNilf` to `NotEmpty` to make it more idiomatic: https://pkg.go.dev/github.com/stretchr/testify/assert#NotEmpty.

Screen Shot 2022-11-03 at 10 37 58 AM

File: persistence/block.go (hunk 6 above)
Comment (Olshansk) [reply persistence/block.go:79]: How about just `IndexTransactions`? Storing could be used with the merkle tree separately

File: persistence/block.go (hunk 3 above)
Comment (Olshansk) [reply persistence/block.go:93]: Leave as is for now. Just need a TODO near the method signature that I'll remove it.

File: persistence/indexer/doc/README.md [hunk 18]
```diff
@@ -1,16 +1,16 @@
 # Transaction Indexer
 
-`txIndexer` implementation uses a `KVStore` (interface) to index the transactions.
+`txIndexer` implementation uses a `KVStore` (interface) to txIndex the transactions.
```
Comment (Olshansk) [inline persistence/indexer/doc/README.md:3]: Is this change intentional? There are no changes in the indexer business logic

**Review by Olshansk — CHANGES_REQUESTED** [review]
@andrewnguyen22 Left a few more minor comments, but only have one point of contention.

> Understood - we should consolidate this terminology. Do you think that needs to be done in this commit?

I don't think we need to fix/consolidate _everything_, but I do think that for the methods & variables introduced in this commit, I'd prefer to use `Proposal` or `Previous` instead of `Latest` where appropriate. I personally confuse `Latest` with `Current` whenever I read it, and am not sure if it's Proposal related data (uncommitted) ore previously committed data.


Screen Shot 2022-11-03 at 10 44 39 AM
Screen Shot 2022-11-03 at 10 42 22 AM

File: persistence/block.go (hunk 6 above)
Comment (andrewnguyen22) [reply persistence/block.go:79]: That makes sense.

**Comment by andrewnguyen22** [comment]
@Olshansk

>  don't think we need to fix/consolidate everything, but I do think that for the methods & variables introduced in this commit, I'd prefer to use Proposal or Previous instead of Latest where appropriate. I personally confuse Latest with Current whenever I read it, and am not sure if it's Proposal related data (uncommitted) ore previously committed data.

https://github.com/pokt-network/pocket/pull/332#discussion_r1013034414

So in this comment, it is communicated that there are `non-proposal` blocks going through this lifecycle. With this being in mind, I'm not sure your recommendation makes sense. I will use Previous then instead

File: persistence/indexer/doc/README.md (hunk 18 above)
Comment (andrewnguyen22) [reply persistence/indexer/doc/README.md:3]: Reverting!

**Comment by Olshansk** [comment]
> WYT about next steps?

Reviewed the PR, and this is the last item before we can merge it in.

The Postgres context is attached to a specific height and is meant to be an "ephemeral context" during the application/proposal lifecycle. What if we just drop the `latest` prefix altogether? We know that it's attached to the context, and whether that context is used for proposal or application depends on the owner of that context.

**Comment by andrewnguyen22** [comment]
@Olshansk Corrected as specified and conflicts resolved - give it another go when you can

**Review by Olshansk — APPROVED** [review]
Solid. 🚢
